### ISSUE TITLE
multiple setLastModified may set CACHED when page needs refresh

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3807]
|<img alt="jmg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jmg reported|
|-|-|
|Trac ID|trac#3807|
|Type|defect|
|Created|2009-04-30 21:56:54Z|
|Branch|https://github.com/twisted/twisted/tree/web-http-setLastModified-cache-3807|

The documentation says:
"If I am called more than once, I ignore attempts to set Last-Modified earlier, only replacing the Last-Modified time if it is to a later value."

But, if you call it more than once, and a call attempts to set last modified earlier than the if-modified-since header it won't change the last-modified time, but it will set the cache flag on the request not sending a body even though it should have been sent.

Attachments:

* [ticket-3807-0.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/507/507c2bde546157d98e47afb0e34ece2c9c7edc44/423953cc289e671857775ae3cf5feb0d620914db.patch) (4709 bytes) - added by donal on 2016-03-08 04:42:59Z - One-line fix + tests + newsfiles

<details><summary>Searchable metadata</summary>

```
trac-id__3807 3807
type__defect defect
reporter__jmg jmg
priority__normal normal
milestone__ 
branch__web_http_setLastModified_cache_3807 web-http-setLastModified-cache-3807
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1241128614000000 1241128614000000
changetime__1467226683127712 1467226683127712
version__None None
owner__donal donal

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jmg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|jmg commented|
|-|-|

Apparently I cannot attach files currently.  See [#3808](https://github.com/twisted/twisted/issues/3808).  Same thing happens when I attempt to attach it to this bug.  Here is the patch inline:
--- http.py.orig        Thu Apr 30 14:40:13 2009
+++ http.py     Thu Apr 30 14:40:32 2009
@@ -794,6 +794,7 @@
         if (not self.lastModified) or (self.lastModified < when):
             self.lastModified = when
 
+        when = self.lastModified
         modified_since = self.getHeader('if-modified-since')
         if modified_since:
             modified_since = stringToDatetime(modified_since.split(';', 1)[0])

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="donal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|donal commented|
|-|-|

The tests in this patch may also adequately address [[#6127](https://github.com/twisted/twisted/issues/6127)]

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

(In [46962]) Branching to web-http-setLastModified-cache-3807.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to donal|
|-|-|

Thanks for the changes. They look good.

Major comments:


1. We should not write tests targeted for a specific ticket. Please consider writing indepedent test for which the description is self contained. 

2. For `test_setLastModified` the docstring informs that ``` and sets the Last-Modified header for the response. ``` but I don't see any assertion in those tests.

3. Same for ``` should return None. Otherwise, CACHED. ``` the test itself never checks for the CACHE value so I would say that the docstring is not valid...or the test is not valid :)

Minor comments:

1. For the test docstring, I prefer to have the important description which change between test described first, so that it easier to read what is specific to this test. There is a lot of duplication in the current docstrings... and I smell they were written starting with a copy/paste.. which I think is a bad practice :)


---------

Please check my comments and reconsider updating/rewriting the test descriptions so that the test scenario is clear and the description also match what is tested :)

Also see https://jml.io/pages/test-docstrings.html

I have sent your branch to buildbot.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

[#6127](https://github.com/twisted/twisted/issues/6127) is a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc removed owner|
|-|-|

See:
https://github.com/twisted/twisted/pull/229

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to donal|
|-|-|

Looks good. Thanks.

Will wait for tests to run and will merge.

Thanks again for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

I have created a PR with my changes https://github.com/twisted/twisted/pull/269

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|GitHub set status to `closed`|
|-|-|

In changeset 1429fbacacb2b10d9c708d6016e532118bf36160
```
#!CommitTicketReference repository="" revision="1429fbacacb2b10d9c708d6016e532118bf36160"
Merge web-http-setLastModified-cache-3807: Fix multiple calls to t.w.http.Request.setLastModified.


Author: donalm
Reviewer: adiroiban
Fixes: #3807
```

**Comment by twisted-trac** [comment]
|[<img alt="donalm's avatar" src="https://avatars.githubusercontent.com/u/876167?s=50" width="50" height="50">](https://github.com/donalm)<a name="note_12"></a>|@donalm commented|
|-|-|

Nice - thanks!
